src/bin/decode-faster-mapped-cuda.cc

```diff
@@ -0,0 +1,185 @@
+// WARRANTIES OR CONDITIONS OF TITLE, FITNESS FOR A PARTICULAR PURPOSE,
+// MERCHANTABLITY OR NON-INFRINGEMENT.
+// See the Apache 2 License for the specific language governing permissions and
+// limitations under the License.
+
+#include <cuda_profiler_api.h>
+
+#include "base/kaldi-common.h"
+#include "util/common-utils.h"
+#include "tree/context-dep.h"
+#include "hmm/transition-model.h"
+#include "fstext/fstext-lib.h"
+#include "decoder/cuda-decoder.h"
+#include "decoder/decodable-matrix.h"
+#include "nnet3/nnet-utils.h"
+#include "base/timer.h"
+#include "lat/kaldi-lattice.h" // for {Compact}LatticeArc
+
+int main(int argc, char *argv[]) {
+  try {
+    using namespace kaldi;
+    typedef kaldi::int32 int32;
+    using fst::SymbolTable;
+    using fst::VectorFst;
+    using fst::Fst;
+    using fst::StdArc;
+
+    const char *usage =
+      "Decode, reading log-likelihoods as matrices\n"
+      " (model is needed only for the integer mappings in its transition-model)\n"
+      "Usage:   decode-faster-mapped-cuda [options] <model-in> <fst-in> "
+      "<loglikes-rspecifier> <words-wspecifier> [<alignments-wspecifier>]\n";
+    ParseOptions po(usage);
+    CudaDecoderConfig decoder_opts;
+    bool allow_partial = true;
+    std::string word_syms_filename;
+
+    decoder_opts.Register(&po);
+    po.Register("allow-partial", &allow_partial,
+                "Produce output even when final state was not reached");
+    po.Register("word-symbol-table", &word_syms_filename,
+                "Symbol table for words [for debug output]");
+
+    po.Read(argc, argv);
+
+    if (po.NumArgs() < 4 || po.NumArgs() > 5) {
+      po.PrintUsage();
+      exit(1);
+    }
+
+    std::string model_in_filename = po.GetArg(1),
+                fst_in_filename = po.GetArg(2),
+                loglikes_rspecifier = po.GetArg(3),
+                words_wspecifier = po.GetArg(4),
+                alignment_wspecifier = po.GetOptArg(5);
+
+    TransitionModel trans_model;
+    ReadKaldiObject(model_in_filename, &trans_model);
+
+    Int32VectorWriter words_writer(words_wspecifier);
+
+    Int32VectorWriter alignment_writer(alignment_wspecifier);
+
+    fst::SymbolTable *word_syms = NULL;
+    if (word_syms_filename != "") {
+      word_syms = fst::SymbolTable::ReadText(word_syms_filename);
+      if (!word_syms)
+        KALDI_ERR << "Could not read symbol table from file " << word_syms_filename;
+    }
+
+    CuDevice::Instantiate().SelectGpuId("yes");
+    CuDevice::Instantiate().AllowMultithreading();
+
+    CudaFst cuda_fst;
+    SequentialBaseFloatMatrixReader loglikes_reader(loglikes_rspecifier);
+
+    // It's important that we initialize decode_fst after loglikes_reader, as it
+    // can prevent crashes on systems installed without enough virtual memory.
+    // It has to do with what happens on UNIX systems if you call fork() on a
+    // large process: the page-table entries are duplicated, which requires a
+    // lot of virtual memory.
```

Comment:
A note to myself for later: I am incredibly skeptical about this. I am wondering if cudaMallocHost has anything to do with it (although I realize that only the page table entries are copied, not the pages themselves, when fork() is called). Not to mention that fork() is not called at any point here.

Also, I'm pretty sure that if you do a *big enough* memory allocation, linux has an option to make a page table entry whose size is 1 GB or soemthing (instead of the regular 8KB, or whatever it is). (I think it has something to do with malloc() actually doing mmap() at large enough byte sizes.)

Either way, I am skeptical about this claim.

Reply:
@galv, that comment was just copied over from another program; I added it a long time ago.  Changing the order did resolve some kind of problem we had.  It might interact with Linux's "overcommit" setting.  It's even possible that the interaction was with a program (mem_killer.pl) we have locally that kills processes of users using un-reserved memory on systems that are experiencing memory pressure.